src/ocrd/resource_manager.py

```diff
@@ -322,20 +320,23 @@ def download(
                 Path('out').mkdir()
                 with pushd_popd('out'):
                     mimetype = guess_media_type(f'../{archive_fname}', fallback='application/octet-stream')
-                    log.info("Extracting %s archive to %s/out" % (mimetype, tempdir))
+                    log.info(f"Extracting {mimetype} archive to {tempdir}/out")
                     if mimetype == 'application/zip':
                         with ZipFile(f'../{archive_fname}', 'r') as zipf:
                             zipf.extractall()
                     elif mimetype in ('application/gzip', 'application/x-xz'):
                         with open_tarfile(f'../{archive_fname}', 'r:*') as tar:
                             tar.extractall()
                     else:
-                        raise RuntimeError("Unable to handle extraction of %s archive %s" % (mimetype, url))
-                    log.info("Copying '%s' from archive to %s" % (path_in_archive, fpath))
+                        raise RuntimeError(f"Unable to handle extraction of {mimetype} archive {url}")
+                    log.info(f"Copying '{path_in_archive}' from archive to {fpath}")
                     if Path(path_in_archive).is_dir():
                         copytree(path_in_archive, str(fpath))
                     else:
                         copy(path_in_archive, str(fpath))
+            if Path(tempdir).exists():
+                log.info(f"Removing temp dir {tempdir}")
+                rmtree(tempdir)
```

Comment:
I don't understand. Shouldn't the cleanup already by done by the __exit__ function of `pushd_popd(tempdir=True)`?

In particular, how else would you make sure that the temporary directory gets deleted when an exception has been raised?

Reply:
I have never heard of this (delayed delete) before. Are you sure this is from builtin (`tempfile`) behaviour? In the tests we might have other mechanisms that do not actually use these mechanisms... (`assets.copy_of_directory` has a `rmtree` of its own, but there may be more)

Using `cleanup` kwarg in `pushd_popd` and elsewhere – sure, give it a try. 

But first we need to be sure what's going on – so we can decisively determine if we do avoid lazy cleanup.

Reply:
> I have never heard of this (delayed delete) before. Are you sure this is from builtin (tempfile) behaviour?

I am not sure. 

> But first we need to be sure what's going on – so we can decisively determine if we do avoid lazy cleanup.

Agree. I will try to volume map the `/tmp` of the `ocrd_all` image to verify.

Reply:
I have removed these lines with [7e5946e](https://github.com/OCR-D/core/pull/1246/commits/7e5946e1cbf65baefcd3601fc4a67dbc35e7a327). Everything works fine with the removal. The issue rather seems when all models are downloaded, then the container runs out of memory which leads to the error in #1245

Reply:
Yes, `ocrd resmgr download *` is not recommendable – we are talking about ~ 10 GB currently.

Reply:
It would be 10GB if it had worked correctly. The current size of all downloadable models is not more than 2GBs with `ocrd resmgr download *` with the `ocrd_all` images.

Reply:
Oh, really? So I guess we have to go fix some of our model URLs. Do you have some respective CI log readily available?

Reply:
@bertsky, here I have executed it again:

```sh
(venv38-core) mm@MM-Notebook:~/repos/core$ docker run -v "/home/mm/Desktop/resources:/usr/local/share" ocrd/all:maximum ocrd resmgr download '*'
16:36:30.257 INFO ocrd.cli.resmgr - Downloading registered resource 'qurator-gt4histocr-1.0' (https://qurator-data.de/calamari-models/GT4HistOCR/2019-12-11T11_10+0100/model.tar.xz)

16:36:32.877 INFO ocrd.resource_manager._download_impl - Downloading https://qurator-data.de/calamari-models/GT4HistOCR/2019-12-11T11_10+0100/model.tar.xz to download.tar.xx
16:40:20.742 INFO ocrd.resource_manager.download - Extracting application/x-xz archive to /tmp/tmp6ykhkm5s/out
16:40:25.396 INFO ocrd.resource_manager.download - Copying '.' from archive to /usr/local/share/ocrd-resources/ocrd-calamari-recognize/qurator-gt4histocr-1.0
16:40:25.461 INFO ocrd.cli.resmgr - Installed resource https://qurator-data.de/calamari-models/GT4HistOCR/2019-12-11T11_10+0100/model.tar.xz under /usr/local/share/ocrd-resources/ocrd-calamari-recognize/qurator-gt4histocr-1.0
16:40:25.461 INFO ocrd.cli.resmgr - Use in parameters as 'qurator-gt4histocr-1.0'
16:40:25.461 INFO ocrd.cli.resmgr - Downloading registered resource 'zpd-fraktur19' (https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_fraktur19-1.tar.gz)

16:40:25.461 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_fraktur19-1.tar.gz to download.tar.xx
16:41:02.709 INFO ocrd.resource_manager.download - Extracting application/gzip archive to /tmp/tmpnrj8yfnc/out
16:41:03.135 INFO ocrd.resource_manager.download - Copying 'c1_fraktur19-1' from archive to /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-fraktur19
16:41:03.199 INFO ocrd.cli.resmgr - Installed resource https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_fraktur19-1.tar.gz under /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-fraktur19
16:41:03.199 INFO ocrd.cli.resmgr - Use in parameters as 'zpd-fraktur19'
16:41:03.199 INFO ocrd.cli.resmgr - Downloading registered resource 'zpd-latin-script-hist-3' (https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_latin-script-hist-3.tar.gz)

16:41:03.199 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_latin-script-hist-3.tar.gz to download.tar.xx
16:41:35.393 INFO ocrd.resource_manager.download - Extracting application/gzip archive to /tmp/tmpx5bympjr/out
16:41:35.827 INFO ocrd.resource_manager.download - Copying 'c1_latin-script-hist-3' from archive to /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-latin-script-hist-3
16:41:35.890 INFO ocrd.cli.resmgr - Installed resource https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_latin-script-hist-3.tar.gz under /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-latin-script-hist-3
16:41:35.890 INFO ocrd.cli.resmgr - Use in parameters as 'zpd-latin-script-hist-3'
16:41:35.890 INFO ocrd.cli.resmgr - Downloading registered resource 'qurator-gt4histocr-1.0' (https://qurator-data.de/calamari-models/GT4HistOCR/2019-12-11T11_10+0100/model.tar.xz)


16:41:35.890 INFO ocrd.cli.resmgr - Directory /usr/local/share/ocrd-resources/ocrd-calamari-recognize/qurator-gt4histocr-1.0 already exists but --overwrite is not set
16:41:35.890 INFO ocrd.cli.resmgr - Use in parameters as 'qurator-gt4histocr-1.0'
16:41:35.890 INFO ocrd.cli.resmgr - Downloading registered resource 'zpd-fraktur19' (https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_fraktur19-1.tar.gz)
16:41:35.891 INFO ocrd.cli.resmgr - Directory /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-fraktur19 already exists but --overwrite is not set

16:41:35.891 INFO ocrd.cli.resmgr - Use in parameters as 'zpd-fraktur19'
16:41:35.891 INFO ocrd.cli.resmgr - Downloading registered resource 'zpd-latin-script-hist-3' (https://github.com/Calamari-OCR/calamari_models_experimental/releases/download/v0.0.1-pre1/c1_latin-script-hist-3.tar.gz)
16:41:35.891 INFO ocrd.cli.resmgr - Directory /usr/local/share/ocrd-resources/ocrd-calamari-recognize/zpd-latin-script-hist-3 already exists but --overwrite is not set
16:41:35.891 INFO ocrd.cli.resmgr - Use in parameters as 'zpd-latin-script-hist-3'
16:41:35.891 INFO ocrd.cli.resmgr - Downloading registered resource 'en-default.pyrnn.gz' (https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz)

16:41:37.072 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz to /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/en-default.pyrnn.gz
16:42:18.674 INFO ocrd.cli.resmgr - Installed resource https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz under /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/en-default.pyrnn.gz
16:42:18.675 INFO ocrd.cli.resmgr - Use in parameters as 'en-default.pyrnn.gz'
16:42:18.675 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur.pyrnn.gz' (https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz)

16:42:18.675 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz to /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz
16:42:36.229 INFO ocrd.cli.resmgr - Installed resource https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz under /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz
16:42:36.229 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur.pyrnn.gz'

16:42:36.229 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur-jze.pyrnn.gz' (https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz)
16:42:36.229 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz to /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur-jze.pyrnn.gz
16:42:38.223 INFO ocrd.cli.resmgr - Installed resource https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz under /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur-jze.pyrnn.gz
16:42:38.224 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur-jze.pyrnn.gz'
16:42:38.224 INFO ocrd.cli.resmgr - Downloading registered resource 'LatinHist.pyrnn.gz' (https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz)

16:42:38.224 INFO ocrd.resource_manager._download_impl - Downloading https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz to /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/LatinHist.pyrnn.gz
16:42:45.272 INFO ocrd.cli.resmgr - Installed resource https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz under /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/LatinHist.pyrnn.gz
16:42:45.272 INFO ocrd.cli.resmgr - Use in parameters as 'LatinHist.pyrnn.gz'
16:42:45.272 INFO ocrd.cli.resmgr - Downloading registered resource 'en-default.pyrnn.gz' (https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz)

16:42:45.273 INFO ocrd.cli.resmgr - File /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/en-default.pyrnn.gz already exists but --overwrite is not set
16:42:45.273 INFO ocrd.cli.resmgr - Use in parameters as 'en-default.pyrnn.gz'
16:42:45.273 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur.pyrnn.gz' (https://github.com/zuphilip/ocropy-models/raw/master/fraktur.pyrnn.gz)

16:42:45.273 INFO ocrd.cli.resmgr - File /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur.pyrnn.gz already exists but --overwrite is not set
16:42:45.273 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur.pyrnn.gz'

16:42:45.273 INFO ocrd.cli.resmgr - Downloading registered resource 'fraktur-jze.pyrnn.gz' (https://github.com/jze/ocropus-model_fraktur/raw/master/fraktur.pyrnn.gz)
16:42:45.273 INFO ocrd.cli.resmgr - File /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/fraktur-jze.pyrnn.gz already exists but --overwrite is not set

16:42:45.273 INFO ocrd.cli.resmgr - Use in parameters as 'fraktur-jze.pyrnn.gz'
16:42:45.273 INFO ocrd.cli.resmgr - Downloading registered resource 'LatinHist.pyrnn.gz' (https://github.com/chreul/OCR_Testdata_EarlyPrintedBooks/raw/master/LatinHist-98000.pyrnn.gz)
16:42:45.273 INFO ocrd.cli.resmgr - File /usr/local/share/ocrd-resources/ocrd-cis-ocropy-recognize/LatinHist.pyrnn.gz already exists but --overwrite is not set
16:42:45.273 INFO ocrd.cli.resmgr - Use in parameters as 'LatinHist.pyrnn.gz'
16:42:45.273 INFO ocrd.cli.resmgr - Downloading registered resource 'default-2021-03-09' (https://qurator-data.de/sbb_binarization/2021-03-09/models.tar.gz)

16:42:47.387 INFO ocrd.resource_manager._download_impl - Downloading https://qurator-data.de/sbb_binarization/2021-03-09/models.tar.gz to download.tar.xx
16:43:55.075 INFO ocrd.resource_manager.download - Extracting application/gzip archive to /tmp/tmp6pwrcsgu/out
16:43:55.822 INFO ocrd.resource_manager.download - Copying 'models' from archive to /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default-2021-03-09
16:43:55.907 INFO ocrd.cli.resmgr - Installed resource https://qurator-data.de/sbb_binarization/2021-03-09/models.tar.gz under /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default-2021-03-09
16:43:55.907 INFO ocrd.cli.resmgr - Use in parameters as 'default-2021-03-09'
16:43:55.907 INFO ocrd.cli.resmgr - Downloading registered resource 'default' (https://qurator-data.de/sbb_binarization/models.tar.gz)

16:43:55.908 INFO ocrd.resource_manager._download_impl - Downloading https://qurator-data.de/sbb_binarization/models.tar.gz to download.tar.xx
16:52:36.212 INFO ocrd.resource_manager.download - Extracting application/gzip archive to /tmp/tmp9ti26rws/out
16:52:44.786 INFO ocrd.resource_manager.download - Copying 'models' from archive to /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default
16:52:45.727 INFO ocrd.cli.resmgr - Installed resource https://qurator-data.de/sbb_binarization/models.tar.gz under /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default
16:52:45.727 INFO ocrd.cli.resmgr - Use in parameters as 'default'
16:52:45.727 INFO ocrd.cli.resmgr - Downloading registered resource 'default-2021-03-09' (https://qurator-data.de/sbb_binarization/2021-03-09/models.tar.gz)

16:52:45.727 INFO ocrd.cli.resmgr - Directory /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default-2021-03-09 already exists but --overwrite is not set
16:52:45.727 INFO ocrd.cli.resmgr - Use in parameters as 'default-2021-03-09'
16:52:45.727 INFO ocrd.cli.resmgr - Downloading registered resource 'default' (https://qurator-data.de/sbb_binarization/models.tar.gz)

16:52:45.727 INFO ocrd.cli.resmgr - Directory /usr/local/share/ocrd-resources/ocrd-sbb-binarize/default already exists but --overwrite is not set
16:52:45.727 INFO ocrd.cli.resmgr - Use in parameters as 'default'

```

```sh
(venv38-core) mm@MM-Notebook:~/repos/core$ du -h ~/Desktop/resources/
92M	/home/mm/Desktop/resources/ocrd-resources/ocrd-calamari-recognize/zpd-latin-script-hist-3
92M	/home/mm/Desktop/resources/ocrd-resources/ocrd-calamari-recognize/qurator-gt4histocr-1.0
89M	/home/mm/Desktop/resources/ocrd-resources/ocrd-calamari-recognize/zpd-fraktur19
272M	/home/mm/Desktop/resources/ocrd-resources/ocrd-calamari-recognize
141M	/home/mm/Desktop/resources/ocrd-resources/ocrd-cis-ocropy-recognize
12K	/home/mm/Desktop/resources/ocrd-resources/ocrd
147M	/home/mm/Desktop/resources/ocrd-resources/ocrd-sbb-binarize/default-2021-03-09
1,8G	/home/mm/Desktop/resources/ocrd-resources/ocrd-sbb-binarize/default
1,9G	/home/mm/Desktop/resources/ocrd-resources/ocrd-sbb-binarize
2,3G	/home/mm/Desktop/resources/ocrd-resources
2,3G	/home/mm/Desktop/resources/

```

Not to forget that it's core 2.64.0 inside that is used, although I have pulled the latest version.

Reply:
That's odd. Many processors are not even attempted, like ocrd-detectron2-segment or ocrd-tesserocr-recognize.

Could it be the case that you have some `resources/ocrd/resources.yml` user database which was generated when you had fewer processors installed?

Reply:
I am using the docker container to extract the models. I didn't volume map any 'resources.yml' separately. The '~/Desktop/resources' I am volume mapping is a freshly created directory. 

I would expect the container to not access my host machine to search for that file under my home directory. There is one under '~/.config/ocrd/resources.yml'

Reply:
But surely, such a user database was at least created by that call. And if you did not run any `list-available`, then that database would be just a mirror of the distributed ocrd/resource_list.yml (hence only those 3 processors).

So we just uncovered another serious bug: initialisation does not search the PATH for ocrd-* executables, only `list-available` does. But without these database entries, `list-installed` never even attempts to look for other executables!

Reply:
I opened https://github.com/OCR-D/core/issues/1251 for dedicated discussion. Not sure if this PR or #1250 or an entirely new PR is the best place for this.